### PR TITLE
Draw discard

### PR DESCRIPTION
**Pull request by Zeimyth**

### CODE REVIEW

File: src/framework/actions/hand.ts [hunk 1]
```diff
@@ -4,6 +4,96 @@ import { GameState } from "../types";
...
+  const P = $.players[$.activePlayer];
+
+  if (P.deck.length > 0) {
+    P.hand = P.hand.concat(P.deck.splice(0, 1));
+  } else if (P.hasShuffledThisTurn === false) {
+    P.hasShuffledThisTurn = true;
+    shuffle($);
+    drawCard($);
+  } else {
+    throw new Error("deck is empty");
```
Comment (Sicilica) [inline src/framework/actions/hand.ts]: From the rules:

>  If you manage to empty your draw pile a second time during your main phase, you can’t draw any more cards during that main phase

The rules don't make it explicitly clear what should happen if you are unable to draw a card, but my assumption would be that you simply don't draw anything (as opposed to it being illegal to do something that would cause you to need to draw a card when your deck is empty). An error here should essentially mean "whatever action you just tried to do, _you can't_"; i.e., this would mean if someone tries to kill an enemy technician when they have no cards in deck, their attack is actually not allowed.

With that in mind, I think we should change this to return `boolean` and have it return whether a card was successfully drawn or not (in case there's a card like "draw a card; if you do, ___"), but this should never error.

File: src/framework/actions/hand.ts [hunk 2]
```diff
@@ -4,6 +4,96 @@ import { GameState } from "../types";
...
+ * server. Active players' clients should not be allowed to manipulate their
+ * draw order, and the contents/order of the draw pile for other players needs
+ * to be unknown.
+ * NOTE: This draws the top card. For abilities that allow drawing an arbitrary
+ * card from the player's deck, a different function should be used.
+ *
+ * @param $
+ * @param pid
+ */
+export const drawCard = ($: GameState): void => {
```
Comment (Sicilica) [inline src/framework/actions/hand.ts]: The active player isn't the only player who can draw cards (technician...). This (and `shuffle`, e.g.) should probably take a player ID, or take a `PlayerState` pointer instead of the whole `GameState`.

File: src/framework/actions/hand.ts [hunk 3]
```diff
@@ -4,6 +4,96 @@ import { GameState } from "../types";
...
+ * @param pid
+ */
+const shuffle = ($: GameState): void => {
+  const P = $.players[$.activePlayer];
+
+  if (P.deck.length > 0) {
+    throw new Error("deck is not empty");
+  }
+
+  // Shamelessly copied from https://link.medium.com/1JmrvTx7Y7
```
Comment (Sicilica) [inline src/framework/actions/hand.ts:41]: Fisher-Yates!

File: src/framework/actions/hand.ts [hunk 4]
```diff
@@ -4,6 +4,96 @@ import { GameState } from "../types";
...
+
+/**
+ *
+ * @param $
+ * @param cid when omitted, the card discarded from the player's hand will be
+ * random
+ */
+export const discardCard = (
+  $: GameState,
+  cid: string|undefined = undefined
```
Comment (Sicilica) [inline src/framework/actions/hand.ts]:
```suggestion
  cid?: string,
```

File: src/tests/basic/hand.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        expect($.players[P1].discard.length).to.equal(0);
+      });
+    });
+
+    describe("discardCard()", () => {
+      it("moves cards from hand to discard pile when discarding", () => {
+        discardCard($, "Nautical Dog");
+
+        expect($.players[P1].hand.length).to.equal(2);
+        expect($.players[P1].hand.indexOf("Nautical Dog")).to.equal(-1);
```
Comment (Sicilica) [inline src/tests/basic/hand.ts]:
```suggestion
        expect($.players[P1].hand.includes("Nautical Dog")).to.equal(false);
```

We typically prefer `includes` over `indexOf` in cases where we don't care about index

File: src/tests/basic/hand.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        expect($.players[P1].discard.indexOf("Nautical Dog")).to.equal(0);
+      });
+
+      it("should only move one copy of the specified card", () => {
+        $.players[P1].hand = [ "Nautical Dog", "Nautical Dog" ];
+
+        discardCard($, "Nautical Dog");
+
+        expect($.players[P1].hand.length).to.equal(1);
+        expect($.players[P1].hand.indexOf("Nautical Dog")).to.equal(0);
```
Comment (Sicilica) [inline src/tests/basic/hand.ts]:
```suggestion
        expect($.players[P1].hand).to.deep.equal(["Nautical Dog"]);
```

**Review by Sicilica — COMMENTED** [review]
Only blocking comments are around some `drawCard` edge cases

File: src/framework/actions/hand.ts (hunk 1 above)
Comment (Zeimyth) [reply src/framework/actions/hand.ts]: I intended the behavior that the rules describe, but for some reason I didn't consider how counter-productive an exception would be (more burden on the caller).

I've made this correction now in the 5th commit.

File: src/framework/actions/hand.ts (hunk 2 above)
Comment (Zeimyth) [reply src/framework/actions/hand.ts]: This has been corrected in the 5th commit.

Comment (Sicilica) [reply src/framework/actions/hand.ts]: I didn't call it out explicitly, but we will want this change in discard as well (need to be able to discard cards from opponent's hand, and discardAll doesn't need it but wouldn't hurt to be more explicit)

**Review by Sicilica — APPROVED** [review]
We should probably pass player for discard funcs as well but otherwise 🚢
